### ISSUE TITLE
CircularNotchedRectangle does't work with FloatingActionButton.extended()

### ISSUE THREAD

**Issue by hereisderek** [flutter/flutter#21650]
I'm not sure is extended fab with notchedappbar covered in material design, but this sure doesn't look right to me
![screenshot_1536601427](https://user-images.githubusercontent.com/839795/45314531-1c4f3c00-b586-11e8-9c64-f29c4d436405.png)

using slightly modified template project:

```dart
@override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(title: new Text(widget.title),),
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new Text(
              'You have pushed the button this many times:',
            ),
            new Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: new FloatingActionButton.extended(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        icon: new Icon(Icons.add),
        label: const Text("label"),
      ),
      floatingActionButtonLocation: FloatingActionButtonLocation.endDocked,
      bottomNavigationBar: BottomAppBar(
        clipBehavior: Clip.antiAlias,
        color: Colors.blue,
        shape: CircularNotchedRectangle(),
        child: Material(
          child: SizedBox(width: double.infinity, height: 80.0,),
          color: Theme.of(context).primaryColor,
        ),
      ), 
    );
  }
```

**Comment by EliasDeuss** [comment]
I had the Same problem but only on iOS. It randomly stopped for me somehow over a update but I don't know what caused it.

**Comment by karanhashedin** [comment]
I had this code written this code and couldn't figure out why I am not getting the notch
  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        automaticallyImplyLeading: false,
        title: getTitle(),
      ),
      body: _getAgendaWidget(),
      floatingActionButton: new FloatingActionButton.extended(
        onPressed: () {},
        icon: new Icon(Icons.add),
        label: const Text("label"),
      ),
      floatingActionButtonLocation: FloatingActionButtonLocation.centerDocked,
      bottomNavigationBar: BottomAppBar(
        shape: CircularNotchedRectangle(),
        child: Material(
          child: SizedBox(
            width: double.infinity,
            height: 60.0,
          ),
          color: Theme.of(context).primaryColor,
        ),
      ),
    );
  }

![screen shot 2018-09-26 at 6 08 42 pm](https://user-images.githubusercontent.com/36027389/46080318-4116f680-c1b7-11e8-8d47-93e3ac78e485.png)

**Comment by juani21** [comment]
same issue

**Comment by zoechi** [comment]
@JarrodCColburn that sounds great 👍 

> Do I need to run tests before opening a PR?

Not sure what you mean by that.
Your PR should include tests that prevents regression.
When you create the PR the whole Flutter test suite is run automatically on the CI infrastructure.
If you are unsure you can just create the PR and wait for feedback.

I assume you saw https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md  
https://gitter.im/flutter/contributors is a dedicated channel to contribution and contributors.

**Comment by JarrodCColburn** [comment]
Thanks. The solution I implemented works technically, but I wonder if CircularNotchedRectangle is the place to implement the fix as I've done. One may consider it a StadiumBorder. Should these classes be named in a similar way to BorderShape class perhaps?

**Comment by zoechi** [comment]
@JarrodCColburn I think it would be better to add the comment (and further discussion) to the PR instead.

**Comment by lucaslcode** [comment]
I also encountered this. I think the best implementation wouldn't require changing between different `NotchedShape`s, or at least animating smoothly, as switching between the extended and regular FAB is quite common.

**Comment by lucaslcode** [comment]
@Hixie I tried using 
```
bottomNavigationBar: BottomAppBar(
  shape: AutomaticNotchedShape(
    Border.all(),
    StadiumBorder(side: BorderSide())
  ),
  child: Container(),
),
```
but it looks like `Border` depends on textDirection as I get a failed assertion that it is null? Should I file a bug for that (I would have thought only `BorderDirectional` does)?

**Comment by Hixie** [comment]
Ah, yeah, we should add support for directionality to NotchedBorder.

**Comment by TheChrisED** [comment]
Is there a workaround for using Border in an `AutomaticNotchedShape`?
I want to achieve this effect:
![Notched Rectangular FAB](https://user-images.githubusercontent.com/9448980/54922458-7901bb00-4f08-11e9-9236-6ac318684070.png)
I tried inheriting from `Border` and overriding all methods that depend on `TextDirection` so that they default to `TextDirection.ltr` like this:
`
@override
  Path getOuterPath(Rect rect, {TextDirection textDirection}) {
    super.getOuterPath(rect, textDirection: textDirection ?? TextDirection.ltr);
  }
`
While this took care of the failed Null Assertion @lucaslcode described, it produced this error:
The following assertion was thrown during paint():
Failed assertion: line 2134: '<optimized out>': is not true.
I then simply copied the whole code from the `Border` into my class see if that would work, but `Border` depends on some package private methods so that produced a compiler error. And it's bad practice anyway, so does anybody have an idea how to do this or do I have to write a rectangular border class myself?

**Comment by Hixie** [comment]
@TheChrisED Please ask questions in flutter-dev or on Gitter or on StackOverflow. Please file new bugs and feature requests as new issues. Adding questions to existing bugs just confuses the existing bug.
That said, see AutomaticNotchedShape.

**Comment by TheChrisED** [comment]
TL;DR
@lucaslcode was pretty close, simply use `RoundedRectangleBorder` with a corner radius of 0 (which is the default) instead of `Border`, because it doesn't depend on `TextDirection` like this:
```dart
bottomNavigationBar: BottomAppBar(
  shape: AutomaticNotchedShape(
    RoundedRectangleBorder(),
    StadiumBorder(side: BorderSide())
  ),
  child: Container(),
),
```
This works for including any shape into a bottom app bar with a notch, simply substitute `StadiumBorder` for the border you'd like to use.

The original question was how an extended FAB can be integrated into a bottom app bar with a notch using `CircularNotchedRectangle`, but like @lucaslcode suggested this could be solved by using `AutomaticNotchedShape`, because the extended FAB is actually an arbitrary shape using a `StadiumBorder`, so using `AutomaticNotchedShape` makes sense here. 
This means integrating a notch for a diamond shape into a bottom app bar is basically the same problem.
That said maybe you misunderstood me @Hixie, I didn't need help writing the whole code, I've already got everything needed (e.g. custom DiamondBorder code), the only thing I needed was the notch.
I thought that my experience in trying to find a solution and failing might give someone else an idea how to solve this, this is why I posted here.
Actually, it gave me another idea how to solve this and it actually worked :)
I still don't understand why `Border` requires a `TextDirection` in the first place, but I'm sure there's a valid reason.

**Comment by TahaTesser** [comment]
Issue still exist 

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel dev, v1.18.0-6.0.pre, on Mac OS X 10.15.4 19E287, locale
    en-GB)
    • Flutter version 1.18.0-6.0.pre at /Users/taha/Code/flutter_dev
    • Framework revision 84c84fb249 (31 hours ago), 2020-04-20 21:35:01 -0400
    • Engine revision 2bde4f0ae4
    • Dart version 2.9.0 (build 2.9.0-1.0.dev a12c36dd97)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.
[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1
[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build
      1.8.0_212-release-1586-b4-5784211)
[✓] VS Code (version 1.44.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.1
[✓] Connected device (4 available)
    • SM M305F   • 32003c30dc19668f • android-arm64  • Android 10 (API 29)
    • macOS      • macOS            • darwin-x64     • Mac OS X 10.15.4 19E287
    • Chrome     • chrome           • web-javascript • Google Chrome
      81.0.4044.113
    • Web Server • web-server       • web-javascript • Flutter Tools
• No issues found!
```

</details>

**Comment by flutternoob** [comment]
@TheChrisED Hi, I was kind of able to achieve the desired effect as per your post. In the 'floatingActionButton' property of the Scaffold Widget, use the following code: 

```
FloatingActionButton(
            backgroundColor: Colors.black,
            child: Icon(Icons.camera_alt, size: 20,),
            shape: BeveledRectangleBorder(
              borderRadius: BorderRadius.all(Radius.circular(70))
            ),
            onPressed: (){},
          ),
```

And, in the BottomAppBar Widget (I am assuming you want to achieve the above mentioned effect using a BottomAppBar Widget), use the following code:

```
ColoredBox(
      color: Colors.transparent,
      child: BottomAppBar(
        color: Colors.white,
        notchMargin: 7.5,
        shape: AutomaticNotchedShape(
          RoundedRectangleBorder(
          ),
          BeveledRectangleBorder(
              borderRadius: BorderRadius.all(Radius.circular(70))
          )
        ),
        child: Container(
          height: 60,
          child: Padding(
            padding: const EdgeInsets.all(10.0),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: <Widget>[
                Icon(Icons.subject),
                Icon(Icons.search)
              ],
            ),
          ),
        ),
      ),
    )
```

I could not figure out which IconData you've used in your design for the first Icon in the Row (in the BottomAppBar), so I used the closest match.

![screenshot](https://user-images.githubusercontent.com/62204476/83357435-e7060580-a389-11ea-8b5a-c145b4a80c59.png)

**Comment by matamune94** [comment]
> I believe I have a fix for this on my fork at [JarrodCColburn@17a0fe6](https://github.com/JarrodCColburn/flutter/commit/17a0fe61269823c7f9e462596b5aedf794d243bc)
> 
> But I'm a little confused by the contributing process? Do I need to run tests before opening a PR?
> 
> ![c21209ba-dc3a-4a4a-a864-764e9c5bb121](https://user-images.githubusercontent.com/16673615/51775869-c60a1200-20bc-11e9-9ba1-107c8da6cb67.jpeg)

how to get curves under the button ?

**Comment by JarrodCColburn** [comment]
I submitted a PR for this #64292

**Comment by TheBiirb** [comment]
@clocksmith Is this technically an issue? Since extended FABs are not circular, why would you even use `CircularNotchedRectangle`? You can just use `AutomaticNotchedShape` and pass in `StadiumBorder()` for the guest's shape, since by default extended FABs are stadium shaped and it should work.

**Comment by pedromassangocode** [comment]
This looks more like building a bad UI intentionally rather than a bug. The docs of `CircularNotchedRectagle` clearly says:

![Screen Shot 2021-01-19 at 14 47 41](https://user-images.githubusercontent.com/67424416/105043054-4aea5500-5a65-11eb-859d-f583f0b3103e.png)


For anyone that wants the `CircularNotchedRectagle` to work with extended FAB or wants a similar support, then please open a new issue requesting for this new capability and making reference to this current issue.
Thank you
